tensorflow/core/platform/default/logging.cc

```diff
@@ -392,5 +345,182 @@ bool LogEveryNSecState::ShouldLog(double seconds) {
   return true;
 }
 
+std::mutex TFLogSinks::instanceMutex_;
+std::atomic<TFLogSinks*> TFLogSinks::instance_ = ATOMIC_VAR_INIT(nullptr);
+
+TFLogSinks& TFLogSinks::Instance() {
+  auto instance = instance_.load(std::memory_order_acquire);
+  if(!instance) {
+    std::lock_guard<std::mutex> lock(instanceMutex_);
+    instance = instance_.load(std::memory_order_relaxed);
+    if(!instance) {
+      instance = new TFLogSinks;
+      instance_.store(instance, std::memory_order_release);
+    }
+  }
+
+  return *instance;
+}
+
+void TFLogSinks::Add(TFLogSink* sink) {
+  if(sink == nullptr) {
+    return;
+  }
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto it = FindSink(sink);
+  if(it == sinks_.end()) {
+    sinks_.emplace_back(sink);
+  }
+}
+
+void TFLogSinks::Remove(TFLogSink* sink) {
+  if(sink == nullptr) {
+    return;
+  }
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto it = FindSink(sink);
+  if(it != sinks_.end()) {
+    sinks_.erase(it);
+  }
+}
+
+void TFLogSinks::Send(const TFLogEntry& entry) {
+  std::lock_guard<std::mutex> lock(mutex_);
+#ifndef NO_DEFAULT_LOGGER
```

Comment:
Let's simplify the code path by registering `TFDefaultLogSink` in `TFLogSinks`'s constructor conditionally.  Then we don't need this macro here.

Reply:
The default logger is not meant to be part of the queue. It's meant to be a fallback if other log sink is present. I can't just add it to the the queue by default as it has quite severe repercussions. These are the 3 other "solutions" (in the order from simplest to most complicated) I can think of if we did as you suggest: 

* the default logger will always be present with no way for a user to remove it
* we'd have to or skip the first logger in the queue if there's more than one (with the first entry always being the default logger).
* we'd have to remove the default logger the first time another log sink is added. This means that we'd either have to check the first entry to see if it's the default logger or introduce a flag that's set the first time a sink is added. This would also required additional logic to re-add the default logger when the last log sink is removed

After considering the above three solutions, I think I chose the simplest (keep the default logger separate from the queue). In complexity is probably about the same as #2 in the list, but the first one introduces a bug as far as I'm concerned.

Reply:
Could you elaborate what you mean by "severe repercussions"?  Also, I'm fine with introducing `TFGetLogSinks()` function that returns all the added loggers if removing the default logger at runtime is important.

Reasons I prefer treating default logger like any other loggers:
- Macro config is essentially a different build so we won't have a test coverage, better to reduce them.
- Enabling/disabling default logger based on other logger status can be confusing, especially when default logger can be added just like any other loggers, `TFAddLogSink(new TFDefaultLogSink());`.

Reply:
What I meant by "severe repercussions"  is increase in complexity and/or loss of functionality versus having the "default logger" be separate. However, I think you suggested a way to have it both ways if you will.

Let me elaborate on the reason why the "default logger" functionality is needed at all. The problem we ran into was that the dynamic loading and initialization of CUDA components produced log messages that we couldn't intercept because that code ran before we could actually register a log sink. This means that you get stdout spam whether you want it or not.

The solution was the "default logger". The idea is that when the default logger is present, TensorFlow works as expected. This means that no logging changes are needed anywhere else: everything works as expected. However, if you need to take full control of logging, you can set the build flag to disable the default logger. This way you can install a log sink after some of the logging has already taken place. You won't lose log messages because they're in the queue and will be displayed with the log sink you've installed.

The issue with not having a default logger and requiring the user to install a logger before they see any log messages is that it would break existing expected behavior.

Just for completeness, I thought that maybe having the `no_default_logger` flag be an environment variable could solve our issue, but it won't. It's the same issue as we already have with logging happening before the user has a chance to install a log sink. As far as I know, there's no way to delay this initialization, since it seems to be triggered by DLL load.

We may have a compromise solution though, thanks to your suggestion:

1. Add the `TFGetLogSinks()` function as you suggested.
2. Still have the `no_default_logger` flag, but now it only controls one line of code: adding the default logger in the constructor, so there's no need for big chunks of code to be enabled/disabled with the preprocessor.

Thoughts?

Reply:
Another thing to think about in the future is to make the "default logger be a "console logger" (or "stdout logger" or something) and move the Android stuff into a separate log sink, which could even be added in a different place and/or context. That whole thing is the most egregious preprocessor include/exclude usage in `logging.cc` and it's just plain ugly. 

That's out of scope for this I think. I modified default/Android logger code as little as I could because I don't have an Android environment to develop against, that's why the build broke a couple of times as I discovered some "copypasta" mistakes I made. It's just a suggestion for future enhancements, a good project for a junior engineer to cut his/her teeth on.

Reply:
- 1.2.3. sounds good.
- Could you add google3/third_party/tensorflow/core/platform/logging_test.cc that tests 1. Adding and removing mock log sink 2. logger queue ?

Reply:
Sure, the test might be lagging behind the other code changes. I'm a novice at best at using Bazel, so the initial setup will be a learning experience.

Reply:
Also, I won't be able to write an equivalent test for Android.

Reply:
Will do